Refine style guide (#11394)
PyTorchLightning/pytorch-lightning: docs/source/starter/style_guide.rst

###########
Style Guide
###########

> A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo or a research project,
> finding a lightning module and knowing exactly where to look to find the things you care about.
> 

Reviewer: ```suggestion
A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo or a research project,
finding a lightning module, and knowing exactly where to look to find the things you care about.

```

###########
Style Guide
###########

A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo or a research project,

> finding a lightning module, and knowing exactly where to look to find the things you care about.

Reviewer 2: ```suggestion
finding a :class:`~pytorch_lightning.core.lightning.LightningModule`, and knowing exactly where to look to find the things you care about.
```

A model is something like a resnet18, RNN, etc.

A system defines how a collection of models interact with each other with user-defined training/evaluation logic. Examples of this are:

* GANs
* Seq2Seq
* BERT
* etc...

A LightningModule can define both a system and a model.

Here's a LightningModule that defines a model. Although, we do not recommend to define a model like in the example.

.. testcode::

    class LitModel(LightningModule):
        def __init__(self):
            super().__init__()
            self.layer_1 = nn.Linear()
            self.layer_2 = nn.Linear()
            self.layer_3 = nn.Linear()

> Here's a LightningModule that defines a system. This is how we recommend.

Reviewer 3: ```suggestion
Here's a LightningModule that defines a system. This structure is what we recommend as a best practice.
```

> Here's a LightningModule that defines a system. This structure is what we recommend as a best practice.

Reviewer: This should come first in the docs since it's the recommended path to follow 

Somewhere here we should emphasize that this distinction improves modularity, which has a slew of benefits (testability, reduced dependencies, easier to refactor)

A model is something like a resnet18, RNN, etc.

A system defines how a collection of models interact with each other with user-defined training/evaluation logic. Examples of this are:

* GANs
* Seq2Seq
* BERT

> * etc...

Reviewer 4: ```suggestion
* etc.
```
I guess it is "etc." or "..." but not both

A model is something like a resnet18, RNN, etc.

A system defines how a collection of models interact with each other with user-defined training/evaluation logic. Examples of this are:

* GANs

> * Seq2Seq

Reviewer 5: The below bullet could be simply deleted.
```
* etc
```

Author: why ? 🤔

Reviewer 5: It just looks unnecessary to me, but if you think it looks nicer as is, let's keep it :)

.. testcode::

    class LitModel(LightningModule):

>         def __init__(self):

Reviewer: "A LightningModule can define both a system and a model" defeats the point above. 
Users should define pure nn.Modules *outside* of the LightningModule. Including the examples with the layers like this isn't how I've seen the vast majority of users interact with Lightning

Author: will update!

Author: It's a choice I'd say... although a system is what we recommend but if it's a very simple model, then we don't enforce it.

Reviewer 5: I agree with @ananthsub.

@rohitgr7 How about explicitly mentioning here that users are discouraged to define a "model" like the example (although users can still write that way)?

.. testcode::

    class LitModel(LightningModule):

>         def __init__(self, encoder: nn.Module = None, decoder: nn.Module = None):

Reviewer: n00b question: why remove the typehint?

Author: we don't usually have typehints in doc examples.

Reviewer 6: yes agreed

>         def __init__(self, encoder, decoder=None):

Reviewer 5: nit
```suggestion
        def __init__(self, encoder, decoder):
```

Forward vs training_step
========================

We recommend using forward for inference/predictions and keeping ``training_step`` independent.

.. code-block:: python

    def forward(self, x):
        embeddings = self.encoder(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        z = self.encoder(x)
        pred = self.decoder(z)
        ...

However, when using DataParallel, you will need to call forward manually:

.. code-block:: python

    def training_step(self, batch, batch_idx):
        x, y = batch
        z = self(x)  # < ---------- instead of self.encoder(x)
        pred = self.decoder(z)

>         ...

Reviewer: should we recommend users to implement forward then to avoid a special carveout for DataParallel? it's not intuitive why DataParallel requires this

Author: DP is a special case worth highlighting. Advising implementing a forward method always might not be good for this section.

Forward vs training_step
========================

We recommend using forward for inference/predictions and keeping ``training_step`` independent.

.. code-block:: python

    def forward(self, x):
        embeddings = self.encoder(x)
        return embeddings

    def training_step(self, batch, batch_idx):
        x, y = batch
        z = self.encoder(x)
        pred = self.decoder(z)
        ...

However, when using DataParallel, you will need to call forward manually:

.. code-block:: python

    def forward(self, x):
        embeddings = self.encoder(x)
        return embeddings

    def training_step(self, batch, batch_idx):
        x, y = batch
        z = self(x)  # < ---------- instead of self.encoder(x)

>         pred = self.decoder(z)

Reviewer 2: a question in general (even though you did not change this): If we need to call `self(x)` instead of `self.encoder(x)`, can we still call `self.decoder(z)` or would this also have to go to `forward`? I don't see a difference there. If there is one, it probably has to be explained here...

Author: yes.. this isn't required anymore. I think.
since batch is already splitter at this point, so calling self(), or self.encoder(x) won't be a problem of any sort.

for DP model is warpped around `model=DataParallel(LightningParallelModule(LightningModule)))`

`LightningParallelModule` takes care of which method to call, while `DataParallel` splits the batch and calls LightningParallelModule's forward which then calls `training_step` while training with the splitted batch.

cc: @akihironitta

Author: this is what I could find from my investigation. But if anyone can confirm this, I'll update this.

Reviewer 2: I think so as well :)

Author: thanks @justusschock for bringing this up :)

****
Data
****

These are best practices for handling data.

Dataloaders
===========

Lightning uses :class:`~torch.utils.data.DataLoader` to handle all the data flow through the system. Whenever you structure dataloaders,
make sure to tune the number of workers for maximum efficiency.

.. warning:: Make sure not to use ``Trainer(strategy="ddp_spawn")`` with ``num_workers>0`` in a DataLoader or you will bottleneck your code.

DataModules
===========

> Lightning introduced :class:`~pytorch_lightning.core.datamodule.LightningDataModule` in v0.9 release. They are designed as a way of decoupling data-related

Reviewer: the version it was introduced in is not so useful imo

Author: will update!

****
Data
****

These are best practices for handling data.

Dataloaders
===========

Lightning uses :class:`~torch.utils.data.DataLoader` to handle all the data flow through the system. Whenever you structure dataloaders,
make sure to tune the number of workers for maximum efficiency.

.. warning:: Make sure not to use ``Trainer(strategy="ddp_spawn")`` with ``num_workers>0`` in a DataLoader or you will bottleneck your code.

DataModules
===========

The :class:`~pytorch_lightning.core.datamodule.LightningDataModule` is designed as a way of decoupling data-related
hooks from the :class:`~pytorch_lightning.core.lightning.LightningModule` so you can develop dataset agnostic models. It makes it easy to hot swap different
datasets with your model, so you can test it and benchmark it across domains. It also makes sharing and reusing the exact data splits and transforms across projects possible.

> The problem with dataloaders is that sharing full datasets is often still challenging because all these questions need to be answered:

Reviewer 6: This does not longer fit well since you have introduced the concept already above. So the impact of asking these questions is lost. Let's make sure the flow of the intro does not break here.

Author: removed the long intro.. I guess the reference is enough.

****
Data
****

These are best practices for handling data.

Dataloaders
===========

Lightning uses :class:`~torch.utils.data.DataLoader` to handle all the data flow through the system. Whenever you structure dataloaders,
make sure to tune the number of workers for maximum efficiency.

.. warning:: Make sure not to use ``Trainer(strategy="ddp_spawn")`` with ``num_workers>0`` in a DataLoader or you will bottleneck your code.

DataModules
===========

The :class:`~pytorch_lightning.core.datamodule.LightningDataModule` is designed as a way of decoupling data-related
hooks from the :class:`~pytorch_lightning.core.lightning.LightningModule` so you can develop dataset agnostic models. It makes it easy to hot swap different
datasets with your model, so you can test it and benchmark it across domains. It also makes sharing and reusing the exact data splits and transforms across projects possible.

> Checkout :ref:`data` document to understand data management within Lightning and its best practices.

Reviewer 6: ```suggestion
Check out :ref:`data` document to understand data management within Lightning and its best practices.
```
Quiz:
Which of the following is a valid git command?

A) git check out
B) git checkout
C) git chickenout

Author: C) :smile: 

oh! man. I swear I checked it on all PRs when you suggested it on one of them :p
I'll include this as a check for the final doc revamp PR.

Reviewer 6: No worries. I'm just making a little fun of git 😄

It's for this reason that we recommend you use datamodules. This is specially important when collaborating because
it will save your team a lot of time as well.

All they need to do is drop a datamodule into a :class:`~pytorch_lightning.trainer.trainer.Trainer` and not worry about what was done to the data.

This is true for both academic and corporate settings where data cleaning and ad-hoc instructions slow down the progress
of iterating through ideas.

Checkout :ref:`data` doc to understand data management within Lightning and it's best practices.

------------

********
Examples
********

Checkout the live examples to get your hands dirty:

> - `Introduction to PyTorch Lightning <https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/mnist-hello-world.html>`_
> - `Introduction to DataModules <https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/datamodules.html>`_

Reviewer 5: Following Jirka's comment in https://github.com/PyTorchLightning/pytorch-lightning/pull/10440#discussion_r773263578, let's use `stable` instead of `latest`.
```suggestion
- `Introduction to PyTorch Lightning <https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/mnist-hello-world.html>`_
- `Introduction to DataModules <https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/datamodules.html>`_
```

It's for this reason that we recommend you use datamodules. This is specially important when collaborating because
it will save your team a lot of time as well.

All they need to do is drop a datamodule into a :class:`~pytorch_lightning.trainer.trainer.Trainer` and not worry about what was done to the data.

This is true for both academic and corporate settings where data cleaning and ad-hoc instructions slow down the progress
of iterating through ideas.

> Checkout :ref:`data` doc to understand data management within Lightning and it's best practices.

Reviewer 6: ```suggestion
Checkout :ref:`data` doc to understand data management within Lightning and its best practices.
```
